src/pcrlock/pcrlock.c

```diff
@@ -0,0 +1,5208 @@
+
+        r = table_print_with_pager(table, arg_json_format_flags, arg_pager_flags, /* show_header= */ true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to output table: %m");
+
+        printf("\n"
+               "%sLegend: H → PCR hash value matches event log%s\n"
+               "%s        R → All event log records for this PCR have a matching component%s\n"
+               "%s        C → No components that couldn't be matched with log records affect this PCR%s\n",
+               ansi_grey(), ansi_normal(), /* less on small screens automatically resets the color after long lines, hence we set it anew for each line */
+               ansi_grey(), ansi_normal(),
+               ansi_grey(), ansi_normal());
+
+        return 0;
+}
+
+static int event_determine_primary_algorithm(EventLog *el) {
+        assert(el);
+
+        if (el->n_algorithms == 0) {
+                /* Nothing loaded to make the decision on? Then pick SHA256 */
+                el->primary_algorithm = TPM2_ALG_SHA256;
+                return 0;
+        }
+
+        for (size_t i = 0; i < el->n_algorithms; i++) {
+                /* If we have SHA256, focus on that that */
+
+                if (el->algorithms[i] == TPM2_ALG_SHA256) {
+                        el->primary_algorithm = el->algorithms[i];
+                        return 0;
+                }
+        }
+
+        /* Otherwise show the "best" (i.e. the one with the highest id value) */
+        el->primary_algorithm = el->algorithms[el->n_algorithms-1];
+        return 0;
+}
+
+static int event_log_load_and_process(EventLog **ret) {
+        _cleanup_(event_log_freep) EventLog *el = NULL;
+        int r;
+
+        el = event_log_new();
+        if (!el)
+                return log_oom();
+
+        r = event_log_add_algorithms_from_environment(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_load(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_read_pcrs(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_calculate_pcrs(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_validate_record_hashes(el);
+        if (r < 0)
+                return r;
+
+        r = event_determine_primary_algorithm(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_load_components(el);
+        if (r < 0)
+                return r;
+
+        r = event_log_map_components(el);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(el);
+        return 0;
+}
+
+static int verb_show_log(int argc, char *argv[], void *userdata) {
+        _cleanup_(event_log_freep) EventLog *el = NULL;
+        int r;
+
+        r = event_log_load_and_process(&el);
+        if (r < 0)
+                return r;
+
+        putchar('\n');
+
+        r = show_log_table(el);
+        if (r < 0)
+                return r;
+
+        putchar('\n');
+
+        r = show_pcr_table(el);
```

Comment:
In json mode, this results in the output of two unconnected arrays:
```
[
        {
                "pcr" : 0,
                "" : "█",
                "pcrname" : "platform-code",
                "event" : "s-crtm-contents",
                "match" : null,
                "sha256" : "e88b6ad84cc5dd8021a58541e8eb2a8a77476759998ea2bc0990a60f89e634be",
                "F/U" : "F",
                "component" : null,
                "description" : "Raw: Boot Guard Measured S-CRTM\\000"
        },
…
]

[
        …
]
```
This is not valid JSON and not valid json-seq, IIUC. I know you like json-seq as the default format… As discussed previously, this doesn't really work in many scenarios. But in particular, I don't think it makes sense for two different types of data in the same output.

---

JSON looks like this:
```
        {
                "pcr" : 1,
                "" : "█",
                "pcrname" : "platform-config",
                "count" : 20,
                "h" : true,
                "r" : false,
                "c" : true,
                "Calculated_sha256" : "c83d19cf5762f1fd7ebcdd6f790afc2235c55a175834ae0ab1c177fd7f8ab0af",
                "Observed_sha256" : "c83d19cf5762f1fd7ebcdd6f790afc2235c55a175834ae0ab1c177fd7f8ab0af"
        },
```
The field with the empty name is strange, and also `h`, `r`, `c` should not be abbreviated in json output.

Also "Calculated_sha256" → "calculatedSHA256" ?